### ISSUE TITLE
Bazel build of router broken in 6.0.5

### ISSUE THREAD

**Issue by alexeagle** [angular/angular#24521]
This commit
https://github.com/angular/angular/commit/d20877bf3fec3b6475c7ebba440366101b272549#diff-c68ccdefc54b5e3375135dddaa34d46c 
causes ngc to produce a summary file 

```
{"__symbol":110,"name":"ɵEmptyOutletComponent","filePath":"./router"},
{"__symbol":111,"name":"ApplicationRef","filePath":"@angular/core"},
{"__symbol":112,"name":"SystemJsNgModuleLoader","filePath":"@angular/core"},
{"__symbol":113,"name":"ɵflatten","filePath":"./router"},
{"__symbol":114,"name":"ɵEmptyOutletComponentNgFactory","filePath":"./router.ngfactory"},
{"__symbol":115,"name":"View_ɵEmptyOutletComponent_0","filePath":"./router.ngfactory"},
{"__symbol":116,"name":"RenderType_ɵEmptyOutletComponent","filePath":"./router.ngfactory"}]}
```

to point to a ngfactory in a way that breaks the compile:
https://github.com/alexeagle/angular-bazel-example/pull/154
```
ERROR: /home/circleci/ng/src/hello-world/BUILD.bazel:13:1: Compiling Angular templates (ngc) //src/hello-world:hello-world failed (Exit 1)

: Error: Could not resolve ./router.ngfactory from /home/circleci/.cache/bazel/_bazel_circleci/9ce5c2144ecf75d11717c0aa41e45a8d/execroot/angular_bazel_example/node_modules/@angular/router/router.d.ts
    at TsCompilerAotCompilerTypeCheckHostAdapter.fromSummaryFileName (/home/packages/compiler-cli/src/transformers/compiler_host.ts:274:13)
    at AotSummaryResolver.fromSummaryFileName (/execroot/angular/packages/compiler/src/aot/summary_resolver.ts:64:22)
    at /execroot/angular/packages/compiler/src/aot/summary_serializer.ts:441:34
    at Array.map (<anonymous>)
    at FromJsonDeserializer.deserialize (/execroot/angular/packages/compiler/src/aot/summary_serializer.ts:439:33)
    at deserializeSummaries (/execroot/angular/packages/compiler/src/aot/summary_serializer.ts:61:23)
    at AotSummaryResolver._loadSummaryFile (/execroot/angular/packages/compiler/src/aot/summary_resolver.ts:118:13)
    at AotSummaryResolver.resolveSummary (/execroot/angular/packages/compiler/src/aot/summary_resolver.ts:73:12)
    at StaticSymbolResolver._resolveSymbolFromSummary (/execroot/angular/packages/compiler/src/aot/static_symbol_resolver.ts:234:42)
    at StaticSymbolResolver.resolveSymbol (/execroot/angular/packages/compiler/src/aot/static_symbol_resolver.ts:85:36)
```

**Comment by alexeagle** [comment]
@alxhub explains: this router class is the first time we have an `@Component` inside the `@angular` packages.

There's a discrepancy between the generated `ngfactory.js`
```
$ grep EmptyOutletComponentNgFactory node_modules/@angular/router/router.ngfactory.js
var RouterModuleNgFactory = i0.ɵcmf(i1.RouterModule, [], function (_l) { return i0.ɵmod([i0.ɵmpd(512, i0.ComponentFactoryResolver, i0.ɵCodegenComponentFactoryResolver, [[8, [ɵEmptyOutletComponentNgFactory]], [3, i0.ComponentFactoryResolver], i0.NgModuleRef]), i0.ɵmpd(1073742336, i1.RouterModule, i1.RouterModule, [[2, i1.ɵangular_packages_router_router_a], [2, i1.Router]])]); });
var ɵEmptyOutletComponentNgFactory = i0.ɵccf("ng-component", i1.ɵEmptyOutletComponent, View_ɵEmptyOutletComponent_Host_0, {}, {}, []);
exports.ɵEmptyOutletComponentNgFactory = ɵEmptyOutletComponentNgFactory;
```

and the `ngfactory.d.ts`

```
$ cat node_modules/@angular/router/router.ngfactory.d.ts 
import * as i0 from '@angular/core';
import * as i1 from '@angular/router';
export declare const RouterModuleNgFactory: i0.NgModuleFactory<i1.RouterModule>;
```

**Comment by alexeagle** [comment]
We figure that Bazel angular usage has *never* worked with components that come from third-party libraries in node_modules.
We are writing deep imports in the user factory files, like `@angular/router/router.ngfactory`, and these deep imports don't work with bazel resolution strategy.

We are in a mixed mode, where we generate commonjs/es5 ngfactory.js files which Rollup won't know how to resolve.

We could probably find some fix for this, but this doesn't look the same as either google3 or current external, so we are adding complexity. Maybe it is better to fix this by moving all the way to Bazel compilation of libraries from source.

**Comment by alexeagle** [comment]
Another possibility to fix the regression in the short-term by moving this component to some other NgModule in the router package so that user code doesn't generate such an import unless they've opted in.

**Comment by alexeagle** [comment]
@gregmagolan will take a look at making Angular source dependency work in angular-bazel-example, which would give us a principled way to avoid these issues.

**Comment by gregmagolan** [comment]
@alexeagle angular-bazel-example with angular build from source is working with the latest router here: https://github.com/bazelbuild/rules_nodejs/pull/240. Looks promising. This would resole this issue since the build no longer uses mixed mode (bazel & ngc) for angular. Although in the example, mixed mode is still needed for `@ngrx`.

**Comment by mhevery** [comment]
@gregmagolan does that mean we can close this issue?

**Comment by gregmagolan** [comment]
@mhevery Seems like the best way to go to build angular from source. Resolves this issue as well as https://github.com/bazelbuild/rules_typescript/issues/192. Once all the related PRs land then this can get closed. Lets leave it open for now.

**Comment by mlc-mlapis** [comment]
@alexeagle @gregmagolan ... using Angular "6.1.0-beta.3" and pure `ngc` compilation produces the following line like `var i3 = require("../../node_modules/@angular/router/router.ngfactory");` (in this example in our factory file `app.module.ngfactory.js`).

The next bundling step (SystemJS or Rollup, ...) for our app creates a not resolvable dependency in the final code.

I suppose that this problem is related how Angular libs are built using Bazel but still don't understand what the comment above means (which / why `@Component`):

> this router class is the first time we have an @Component inside the @angular packages.

and mainly what is the solution to allow correctly compile an app using ngc. Can you give us some more explanation?

**Comment by johnbendi** [comment]
@mlc-mlapis what @alexeagle means by `@Component...` is that the `@angular/router` package didn't previously include that in versions on or prior to `6.0.4` but now it's inclusion is making your build to fail.

**Comment by alexeagle** [comment]
`EmptyOutletComponent` was introduced to fix some other issue - it just so happens no Angular package ever had a Component before. So Bazel users are still stuck on Angular 6.0.4 for now.
Update on this issue: Greg has all the moving parts to switch to a source dependency on Angular, like we do internally at Google, but it's taking some time to land this work.

**Comment by alexeagle** [comment]
No, 6.1.0 is in RC and will release this week or next (?) so I don't think this will make it.
I suppose we should reconsider rolling back the EmptyOutletComponent, or somehow excluding that from Bazel builds. @jasonaden any ideas?

**Comment by jelbourn** [comment]
It looks like this would also prevent Angular Material from using Angular 6.1 (and thus TypeScript 2.9), since it would break what Bazel stuff we do have working.

**Comment by gregmagolan** [comment]
Example of building angular from source to resolve this issue is here: https://github.com/alexeagle/angular-bazel-example/pull/160

That still depends on landing https://github.com/bazelbuild/rules_typescript/pull/223 in rules_typescript which includes a work-around which either needs to be finalized or a principled fix in angular/angular put together

**Comment by danwulff** [comment]
It's very annoying that Angular projects that utilize Material are unable to use TS 2.9 because of this. Do we have an ETA on when this will get fixed? @alxhub

**Comment by gregmagolan** [comment]
Latest angular can now be build with bazel with angular being built from source. https://github.com/alexeagle/angular-bazel-example/pull/160 was merged in and https://github.com/alexeagle/angular-bazel-example is now building Angular 6.1.2.

@danwulff: @jelbourn is working on updating Angular Material as well.

**Comment by gregmagolan** [comment]
@mlc-mlapis You'll need to make some changes to your Bazel build configuration with Angular 6.1.2. This guide should help https://github.com/alexeagle/angular-bazel-example/wiki/Upgrading-Angular-Bazel-build-to-Angular-6.2.1. Feel free to post issues to https://github.com/alexeagle/angular-bazel-example if you run into other issues.

**Comment by mlc-mlapis** [comment]
@gregmagolan ... not sure if we understand each other but we are using just pure `ngc` compiler for project AOT compilation, nothing else (sure on the base of installed NPM @angular packages). So no direct Bazel configuration or so ...

**Comment by gregmagolan** [comment]
@mlc-mlapis Oh I see. That would be a different issue from this one. This issue is specific to building Angular with Bazel. Please file a new issue.

**Comment by mlc-mlapis** [comment]
@gregmagolan ... Ok, I'll file it. But I think that the main origin of the problem is the same ... the existence of the new code:
```
@Component({template: `<router-outlet></router-outlet>`})
export class EmptyOutletComponent {
}
```
in Router package.

**Comment by mlc-mlapis** [comment]
@gregmagolan ... I checked the case once more and we finally found out a way how to configure our build path to accept `@Component` used even inside Angular packages (because in such cases .ngfactory.js files are produced by `ngc` compiler). So I won't file a new issue at all.
